### PR TITLE
Seed data for the Database and Identity

### PR DESCRIPTION
**Pull request by leonidbogin**

### CODE REVIEW

**Comment by leonidbogin** [comment]
GitHub didn't list all files.
image
image
image

**Review by maximsan — APPROVED** [review]
Looks good.

File: DIMS-Core.Database/DIMS-Core.Database/SeedData/PostDeploy.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ Example:      :setvar TableName MyTable                            
+               SELECT * FROM [$(TableName)]                    
+--------------------------------------------------------------------------------------
+*/
+USE [DIMSCore]
+GO
+:r .\dbo.Direction.Table.sql
+Go
+:r .\dbo.TaskState.Table.sql
+Go
```
[the remark this answers is not in the document]
Comment (leonidbogin) [reply DIMS-Core.Database/DIMS-Core.Database/SeedData/PostDeploy.sql:17]: Create a Post-Deployment Script. This script, when publishing, execute two scripts that will add the required data to the database.
[More about SeedData for VS Database Project.](https://keyholesoftware.com/2016/12/12/creating-a-sql-database-project-for-isolated-development/)

File: DIMS-Core/DIMS-Core.Identity/Entities/Role.cs [hunk 2]
```diff
@@ -4,5 +4,10 @@ namespace DIMS_Core.Identity.Entities
 {
     public class Role : IdentityRole<int>
     {
+        public Role(string name)
+        {
+            Name = name;
+            NormalizedName = name;
```
[the remark this answers is not in the document]
Comment (leonidbogin) [reply DIMS-Core/DIMS-Core.Identity/Entities/Role.cs:10]: This is a field from a class Microsoft.AspNetCore.Identity.IdentityRole.

Fixed by calling the parent constructor.
```suggestion
        public Role(string name) : base(name)
        {

        }
```

File: DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                await roleManager.CreateAsync(new Role("admin"));
+            }
+            if (await roleManager.FindByNameAsync("mentor") is null)
+            {
+                await roleManager.CreateAsync(new Role("mentor"));
+            }
+            if (await roleManager.FindByNameAsync("member") is null)
+            {
+                await roleManager.CreateAsync(new Role("member"));
+            }
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs:25]: Could you move this repeatable action in method?

File: DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                {
+                    Email = defaultMentorMail, 
+                    UserName = defaultMentorMail
+                };
+                var result = await userManager.CreateAsync(mentor, defaultMentorPassword);
+                if (result.Succeeded)
+                {
+                    await userManager.AddToRoleAsync(mentor, "mentor");
+                }
+            }
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs:59]: Need to move creating of user in action because it is repeated

**Review by PushMeNow — CHANGES_REQUESTED** [review]
Hi.
I did some comments.

File: DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs (hunk 3 above)
Comment (leonidbogin) [reply DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs]:
```suggestion
        private static async Task CheckAndCreateRole(RoleManager<Role> roleManager, string name)
        {
            if (await roleManager.FindByNameAsync(name) is null)
            {
                await roleManager.CreateAsync(new Role(name));
            }
        }
```

File: DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs (hunk 4 above)
Comment (leonidbogin) [reply DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs]:
```suggestion
        private static async Task CheckAndCreateUser(UserManager<User> userManager, string email, string password, 
string role)
        {
            if (await userManager.FindByNameAsync(email) is null)
            {
                var user = new User()
                {
                    Email = email,
                    UserName = email
                };
                var result = await userManager.CreateAsync(user, password);
                if (result.Succeeded)
                {
                    await userManager.AddToRoleAsync(user, role);
                }
            }
        }
```

File: DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs (hunk 3 above)
Comment (leonidbogin) [reply DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs]: https://github.com/Dev-incubator/DIMS-Core/pull/25/commits/f736bf5086948ce66479390c5879e23d11450480

File: DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs (hunk 4 above)
Comment (leonidbogin) [reply DIMS-Core/DIMS-Core.Identity/Initialize/IdentityInitializer.cs]: https://github.com/Dev-incubator/DIMS-Core/pull/25/commits/ebdc77a4583bb0370cc22b7d9c2a2fea1ad40cf6

**Review by PushMeNow — APPROVED** [review]
Approve

**Pull request merged**